webauthn-authenticator-rs/examples/key_manager/main.rs

```diff
@@ -90,6 +90,9 @@ pub enum Opt {
     /// Gets information about biometric authentication on the device.
     BioInfo,
     /// Enrolls a fingerprint on the device.
+    ///
+    /// Note: you must set a PIN on the device before you can enroll any
```

Comment:
Hmm, actually, this isn't actually true for FIDO v2.1 authenticators which support `alwaysUv`, I'll need to fix that.

Also, `clap` only shows the second sentence for `--help <command>`, not `help <command>`.

Reply:
The plot thickens: according to [the CTAP 2.1 spec](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#enrollingFingerprint), when receiving an enrolment request, the first step is:

> If `pinUvAuthParam` is missing from the input map, end the operation by returning `CTAP2_ERR_PUAT_REQUIRED`.

On the Yubikey Bio in CTAP 2.1 mode, if you try to enrol a fingerprint on a factory-reset (ie: `alwaysUv = true`), regardless of whether you set a PIN, you get rejected if you don't send `pinUvAuthParam`:

```
2023-01-26T02:53:12.353690Z TRACE webauthn_authenticator_rs::ctap2::commands: Sending: BioEnrollmentRequest { modality: Some(Fingerprint), sub_command: Some(1), sub_command_params: Some({Integer(3): Integer(30000)}), pin_uv_protocol: None, pin_uv_auth_param: None, get_modality: false }
2023-01-26T02:53:12.353749Z TRACE webauthn_authenticator_rs::ctap2::commands: CBOR: cmd=9, cbor=Ok(Map({Integer(1): Integer(1), Integer(2): Integer(1), Integer(3): Map({Integer(3): Integer(30000)})}))
2023-01-26T02:53:12.353784Z TRACE webauthn_authenticator_rs::usb: >>> [00, cb, 25, 1b, 7a, 90, 00, 0c, 09, a3, 01, 01, 02, 01, 03, a1, 03, 19, 75, 30, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T02:53:12.357487Z TRACE webauthn_authenticator_rs::usb: <<< [cb, 25, 1b, 7a, 90, 00, 01, 36, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T02:53:12.357604Z ERROR webauthn_authenticator_rs::usb: Ctap error: Ctap(Ctap2PUATRequired)
```

However, if you disable `alwaysUv` (`toggle-always-uv`), the Yubikey Bio seems to be fine with enrolling fingerprints when there's no set PIN:

```
2023-01-26T02:49:08.139517Z TRACE webauthn_authenticator_rs::ctap2::commands: Sending: BioEnrollmentRequest { modality: Some(Fingerprint), sub_command: Some(1), sub_command_params: Some({Integer(3): Integer(30000)}), pin_uv_protocol: None, pin_uv_auth_param: None, get_modality: false }
2023-01-26T02:49:08.139576Z TRACE webauthn_authenticator_rs::ctap2::commands: CBOR: cmd=9, cbor=Ok(Map({Integer(1): Integer(1), Integer(2): Integer(1), Integer(3): Map({Integer(3): Integer(30000)})}))
2023-01-26T02:49:08.139610Z TRACE webauthn_authenticator_rs::usb: >>> [00, 4c, 49, ba, 75, 90, 00, 0c, 09, a3, 01, 01, 02, 01, 03, a1, 03, 19, 75, 30, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T02:49:08.447642Z TRACE webauthn_authenticator_rs::usb: <<< [4c, 49, ba, 75, bb, 00, 01, 02, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T02:49:08.447777Z TRACE webauthn_authenticator_rs::usb: waiting for UserPresenceNeeded
[...]
```

You can even enrol _additional_ fingerprints without authentication of any kind.

If you re-enable `alwaysUv` (note: [toggling `alwaysUv` **does not** require authentication][1]), then the authenticator goes back to preventing enrolment of fingerprints without authentication.

You also can't acquire `BIO_ENROLLMENT` permissions with UV auth:

```
2023-01-26T03:00:23.843222Z TRACE webauthn_authenticator_rs::ctap2::commands: Sending: ClientPinRequest { pin_uv_protocol: Some(2), sub_command: GetPinUvAuthTokenUsingUvWithPermissions, key_agreement: Some(COSEKey { type_: PinUvProtocol, key: EC_EC2(COSEEC2Key { curve: SECP256R1, x: Base64UrlSafeData([80, 213, 125, 160, 9, 232, 55, 120, 152, 97, 189, 16, 71, 206, 33, 133, 37, 83, 217, 201, 166, 74, 83, 242, 34, 127, 243, 128, 71, 129, 113, 188]), y: Base64UrlSafeData([56, 103, 232, 229, 35, 250, 106, 4, 45, 14, 123, 131, 53, 1, 74, 220, 169, 20, 14, 233, 117, 33, 151, 37, 195, 182, 244, 235, 75, 9, 33, 190]) }) }), pin_uv_auth_param: None, new_pin_enc: None, pin_hash_enc: None, permissions: BIO_ENROLLMENT, rp_id: None }
2023-01-26T03:00:23.845520Z TRACE webauthn_authenticator_rs::ctap2::commands: CBOR: cmd=6, cbor=Ok(Map({Integer(1): Integer(2), Integer(2): Integer(6), Integer(3): Map({Integer(1): Integer(2), Integer(3): Integer(-25), Integer(-1): Integer(1), Integer(-2): Bytes([80, 213, 125, 160, 9, 232, 55, 120, 152, 97, 189, 16, 71, 206, 33, 133, 37, 83, 217, 201, 166, 74, 83, 242, 34, 127, 243, 128, 71, 129, 113, 188]), Integer(-3): Bytes([56, 103, 232, 229, 35, 250, 106, 4, 45, 14, 123, 131, 53, 1, 74, 220, 169, 20, 14, 233, 117, 33, 151, 37, 195, 182, 244, 235, 75, 9, 33, 190])}), Integer(9): Integer(8)}))
2023-01-26T03:00:23.845589Z TRACE webauthn_authenticator_rs::usb: >>> [00, f6, 87, 25, bb, 90, 00, 57, 06, a4, 01, 02, 02, 06, 03, a5, 01, 02, 03, 38, 18, 20, 01, 21, 58, 20, 50, d5, 7d, a0, 09, e8, 37, 78, 98, 61, bd, 10, 47, ce, 21, 85, 25, 53, d9, c9, a6, 4a, 53, f2, 22, 7f, f3, 80, 47, 81, 71, bc, 22, 58, 20, 38, 67, e8, e5]
2023-01-26T03:00:23.848503Z TRACE webauthn_authenticator_rs::usb: >>> [00, f6, 87, 25, bb, 00, 23, fa, 6a, 04, 2d, 0e, 7b, 83, 35, 01, 4a, dc, a9, 14, 0e, e9, 75, 21, 97, 25, c3, b6, f4, eb, 4b, 09, 21, be, 09, 08, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T03:00:23.852392Z TRACE webauthn_authenticator_rs::usb: <<< [f6, 87, 25, bb, 90, 00, 01, 40, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]
2023-01-26T03:00:23.852479Z ERROR webauthn_authenticator_rs::usb: Ctap error: Ctap(Ctap2UnauthorizedPermission)
```

I still need to test MakeCredential - I think there is a bug with the checks.

The Feitian BioPass seems to work correctly and refuse to enrol fingerprints until a PIN is set.

[1]: https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#toggle-alwaysUv

Reply:
I'm going to ignore the buggy behaviour in the Yubikeys for now, because they're not blocking for this PR -- this will just attempt to do the right thing and not push the boundaries. This may be something to look at after I have a poke at conformance testing again.

One _could_ authenticate with UV or PIN, but this still prefers UV, even if UV lock-out is in effect. The changes in this PR aren't a complete solution (lock-outs still aren't handled), but it's still _better_.

But to answer the original question:

> Is pin-set a flag in getInfo that we can use to have the code pre-warn or error on this? Or is there a unique error code from the authenticator that it raises in this case?

Authenticators which support PIN auth and it is not configured report `clientPin: false`.

If you try to enroll a fingerprint on a key with no PIN set, it'll return `UserVerificationRequired` without actually trying to enroll anything.